### ISSUE TITLE
Add support for printing CQMs

### ISSUE THREAD

**Issue by arcondello** [dwavesystems/dimod#1188]
It would be useful to have a string representation for CQM models.

**Proposed Solution**
The most obvious thing would be something like
```python
def __str__(self):
    return self.to_lp_file().read()
```
(see https://github.com/dwavesystems/dimod/pull/1052) or
```python
def __str__(self):
    return dimod.lp.dumps(self)
```
assuming that we add `dump` and `dumps` to go with https://github.com/dwavesystems/dimod/pull/1182

LP files are fairly standard and pretty clear to read. However, there are two issues
1) LP files do not handle SPIN variables, but we do. This is fairly easily handled by adding a `Spin` section.
2) LP files have a weird rule where in the objective, quadratic terms need to be bounded by `[ ... ]/2`. We could either keep that quirk, thereby keeping valid LP files, or ignore it for readability.

**Comment by arcondello** [comment]
**Different Proposed Solution**

Something LP-like, but a bit less in the uncanny valley might be
```python
>>> cqm = dimod.CQM()
>>> a, b = dimod.Reals('ab')
>>> i, j = dimod.Integers('ij'
>>> cqm.set_objective(i + 2*j + i*j)
>>> cqm.add_constraint(a + b <= 5, label='hello')
>>> print(cqm)
Minimize
  Integer('i') + 2 * Integer('j') + Integer('i') * Integer('j')
Subject To
  'hello': Real('a') + Real('b') <= 5
```
At least in this case we dodge problem (3) from above by just doing `f'Real({v.variables[0]!r})'`

**Comment by randomir** [comment]
Speaking of `repr`, I'd print only a few significant stats (num vars, constraints, types, etc), instead going for full-eval-ready repr we have on `BQM`.

**Comment by arcondello** [comment]
I don't think we can do the "Variables" section in full generality because `Binary(0) + Binary('0')` is a perfectly valid model and it wouldn't be obvious how to cast them to intermediate labels.
So I think that brings us to
```
Objective
    Binary('i') + 2*Real('j') +  Binary('i')*Real('j')
Constraints:
  'hello': Integer('a') + Integer('b') <= 5
Bounds
  -5 <= Integer('a') <= 5
```
it's a bit verbose but is at least unambiguous.
